### PR TITLE
[improvement](create tablet) backend create tablet round robin among …

### PR DESCRIPTION
**Pull request by deardeng**
…disks

## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/storage_engine.cpp [hunk 1]
```diff
@@ -448,45 +449,85 @@ Status StorageEngine::set_cluster_id(int32_t cluster_id) {
...
+StorageEngine::Disk_remaining_level get_available_level(double disk_usage_percent) {
+    assert(disk_usage_percent >= 0 && disk_usage_percent <= 1);
+    if (disk_usage_percent < 0.7) {
+        return StorageEngine::Disk_remaining_level::LOW;
+    } else if (disk_usage_percent < 0.85) {
+        return StorageEngine::Disk_remaining_level::MID;
+    } else {
+        return StorageEngine::Disk_remaining_level::HIGH;
+    }
+    return StorageEngine::Disk_remaining_level::LOW;
```
Comment (yujun777) [inline be/src/olap/storage_engine.cpp:461]: delete unreachable code ?

File: be/src/olap/storage_engine.cpp [hunk 2]
```diff
@@ -448,45 +449,85 @@ Status StorageEngine::set_cluster_id(int32_t cluster_id) {
     return Status::OK();
 }
 
+StorageEngine::Disk_remaining_level get_available_level(double disk_usage_percent) {
+    assert(disk_usage_percent >= 0 && disk_usage_percent <= 1);
```
Comment (yujun777) [inline be/src/olap/storage_engine.cpp:453]: when be restart, and fe send create rpc to be,  will disk_usage_percent < 0 ?

File: be/src/olap/storage_engine.cpp [hunk 3]
```diff
@@ -448,45 +449,85 @@ Status StorageEngine::set_cluster_id(int32_t cluster_id) {
...
                 if ((_available_storage_medium_type_count == 1 ||
-                     it.second->storage_medium() == storage_medium) &&
-                    !it.second->reach_capacity_limit(0)) {
-                    stores.push_back(it.second);
+                    data_dir->storage_medium() == storage_medium) && 
+                    !data_dir->reach_capacity_limit(0)) {
+                    size_t disk_available = data_dir->disk_available();
+                    DirInfo dir_info;
+                    dir_info.data_dir = data_dir;
+                    dir_info.available_level = get_available_level(disk_available);
```
Comment (yujun777) [inline be/src/olap/storage_engine.cpp:504]: disk_available is the free capacity in bytes in range [0, +00).   but get_available_level's input arg is disk usage in [0, 1]

**Review by yujun777 — COMMENTED** [review]
need update

File: be/src/olap/storage_engine.cpp [hunk 4]
```diff
@@ -448,45 +449,85 @@
...
+}
+
 std::vector<DataDir*> StorageEngine::get_stores_for_create_tablet(
         TStorageMedium::type storage_medium) {
+    struct DirInfo {
+        DataDir* data_dir;
+
+        StorageEngine::Disk_remaining_level available_level;
+
+        bool operator<(const DirInfo& other) {
```
[the remark this answers is not in the document]
Comment (yujun777) [reply be/src/olap/storage_engine.cpp:471]: use ‘bool operator<(const DirInfo& other) const’.   
if missing the last const, clang will compile failed

**Review by yujun777 — COMMENTED** [review]
update

File: be/src/olap/storage_engine.h [hunk 5]
```diff
@@ -492,7 +496,65 @@ class StorageEngine {
...
+        int64 idx = 0;
+    };
+
+    CreateTabletIdxCache(size_t capacity)
+            : LRUCachePolicy(CachePolicy::CacheType::CREATE_TABLET_RR_IDX_CACHE, capacity,
+                             LRUCacheType::NUMBER,
+                             /*stale_sweep_time_s*/ 30 * 60) {}
+
+private:
+    static constexpr char DELIMITER = '-';
```
Comment (cambyzju) [inline be/src/olap/storage_engine.h:557]: unused variable `DELIMITER`

File: be/src/olap/storage_engine.cpp [hunk 6]
```diff
@@ -448,45 +450,89 @@ Status StorageEngine::set_cluster_id(int32_t cluster_id) {
...
+            _create_tablet_idx_lru_cache->set_index(key, init_num);
+        } else {
+            int64_t next_index = curr_index + 1;
+            // int64 overflow
+            if (next_index < 0) {
+                curr_index = 0;
+            }
+            _create_tablet_idx_lru_cache->set_index(key, next_index);
+        }
+
```
Comment (yujun777) [inline be/src/olap/storage_engine.cpp:500]:
```
 curr_index = _create_tablet_idx_lru_cache->get_index(key);
        // -1, lru can't find key
        if (curr_index == -1) {
            curr_index = std::max(0, _last_use_index[medium] + 1);
        }
        _last_use_index[medium] = curr_index;
       _create_tablet_idx_lru_cache->set_index(key,  std::max(0, curr_index + 1));
```

File: be/src/olap/data_dir.cpp [hunk 7]
```diff
@@ -891,6 +891,10 @@ void DataDir::update_remote_data_size(int64_t size) {
     disks_remote_used_capacity->set_value(size);
 }
 
+size_t DataDir::disk_available() const {
+    return _available_bytes;
+}
```
Comment (dataroaring) [inline be/src/olap/data_dir.cpp:896]: available_bytes() is a better name.

File: be/src/olap/storage_engine.cpp [hunk 8]
```diff
@@ -448,45 +450,89 @@ Status StorageEngine::set_cluster_id(int32_t cluster_id) {
...
+    } else if (disk_usage_percent < 0.85) {
+        return StorageEngine::Disk_remaining_level::MID;
+    }
+    return StorageEngine::Disk_remaining_level::HIGH;
+}
+
 std::vector<DataDir*> StorageEngine::get_stores_for_create_tablet(
-        TStorageMedium::type storage_medium) {
+        int64 partition_id, TStorageMedium::type storage_medium) {
+    struct DirInfo {
```
Comment (dataroaring) [inline be/src/olap/storage_engine.cpp:465]: A too long function.

**Review by dataroaring — CHANGES_REQUESTED** [review]

**Review by yujun777 — APPROVED** [review]
LGTM

File: be/src/olap/storage_engine.cpp (hunk 8 above)
Comment (deardeng) [reply be/src/olap/storage_engine.cpp]: fixed

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Comment by deardeng** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**